### PR TITLE
fix(git): carry extended attributes through a cross-device archive (#2919)

### PR DESCRIPTION
**Pull request by sachiniyer**
The last unchecked box on #2919. Claimed on the issue before starting, with the design
call stated up front so it could be vetoed cheaply rather than in review.

Refs #2919 — **not `Closes`**: symlink timestamps is claimed by two other sessions, and
closing the epic out from under them would lose that item.

## What was lost

`rename(2)` keeps extended attributes for free. The copy reproduced only what it was
written to reproduce, so every namespace was dropped — and which filesystem `$AF_HOME`
sits on decided whether a restored worktree kept them.

Not decoration:

- `system.posix_acl_access` / `_default` carry named-user and named-group grants. A
  **directory's default ACL** vanishing is the surprising direction: files created in the
  restored tree afterwards inherit plain umask permissions, which can be **wider** than
  the policy that was archived.
- `security.capability` — a vendored helper silently loses the capability it needs to bind.
- `security.selinux` — a tree comes back mislabeled on an enforcing box.

## Design call (as claimed on the issue)

**Descriptor-anchored**, like the rest of the copier: `Fgetxattr`/`Flistxattr`/`Fsetxattr`
take the descriptors the walk already validated, so no path is re-derived from an fd and
the name-swap race the copier exists to avoid stays avoided. Both platforms have the F*
forms — worth noting my first grep looked in the wrong files and found only the path-based
ones, which would have forced either `/proc/self/fd` or a raw syscall.

**Ordering is load-bearing**, and it is why this sits between the mode and the times: the
access ACL and the mode bits are one state seen two ways. Setting `system.posix_acl_access`
rewrites the mode, and `chmod` rewrites the ACL mask — so an ACL applied *before* the mode
is silently undone. `setxattr` moves ctime only, so the times still land last.

**Failure policy, per-attribute and deliberately asymmetric.** An archive that refuses to
run is a worse outcome than one that could not store a label — but only while the loss is
*logged* rather than silent, which is #2919's actual complaint:

| condition | behaviour |
| --- | --- |
| destination fs holds no xattrs (`EOPNOTSUPP`) | warn **once**, stop trying |
| one attribute refused (`EPERM`/`EACCES`) | warn naming it, keep going |
| anything else (`E2BIG`, `ENOSPC`, …) | **fail the copy** — errors, not policy |

`trusted.*` needs no special case: an unprivileged lister never sees it, so it never
appears in the list to attempt.

## Verification

`xattr.file` / `xattr.dir` come off `knownCrossDeviceDivergence`, so the guard proves this
rather than this description asserting it. **Watched it fail first** with the listing
stubbed out:

```
xattr.dir  diverges ... source=dirvalue  rename=dirvalue  copy=MISSING
xattr.file diverges ... source=filevalue rename=filevalue copy=MISSING
```

Two new tests cover what a single-attribute probe cannot:

- several attributes on one node, and the **names round-trip as a set**, so an extra or
  missing one fails here rather than only when someone asks for the one that vanished;
- a present attribute whose value is **empty** — meaningful, and exactly what a "size 0
  means absent" reading would silently drop;
- a node with **no** attributes, so an empty list is not read as an error.

The support probe **writes to the source tree** rather than inspecting a tree built later:
asking the wrong tree is how the fidelity guard once dropped its xattr comparison entirely
while still reporting a pass, which that test's own comments record.

**What the guard does not cover, stated rather than implied:** an unprivileged fixture
cannot create `security.capability` or a SELinux label, so the privileged namespaces are
exercised by the policy code and not by a test. That limit is in the code comment and here.

Local checks per the current rule — `gofmt -l .` clean, `go build ./...` ok (plus
`GOOS=darwin go build ./session/...`, since this is syscall code), `golangci-lint --fast`
clean, `scripts/lint-file-length.sh` passed, `go test ./session/git/` green in full (23s).
No `deadcode` locally, no containerized suites, nothing against `./daemon/...` or
`./app/...` — CI runs those, including the macOS leg this needs.

Co-Authored-By: Captain Claude <noreply@anthropic.com>

### CODE REVIEW

**Comment by sachiniyer** [comment]
My #2991 was the duplicate here and is closed — this is the right one to keep, and
the errno handling and ERANGE retries in it are better than what I had. But before
mine closed, Codex found four defects in it, and **one of them is present in this
branch too**. I reproduced it against `pr/2993` rather than reading the diff:

```
--- FAIL: TestProbeReadOnlyXattrOn2993
    LOST xattr on ro.txt:  no data available
    LOST xattr on ro-dir:  no data available
```

**A checked-in `0444` file and a `0555` directory lose every extended attribute,
silently.** `openat(2)`/`preserveSourceMode` leave the destination unwritable
before `copySourceXattrs` runs, ordinary namespaces need WRITE permission, so
`Fsetxattr` returns `EACCES` — and the `case errors.Is(err, unix.EPERM),
errors.Is(err, unix.EACCES): continue` arm swallows it as though it were a
privileged namespace.

That arm is right for `security.capability`. It is what makes this invisible for
an ordinary read-only file, which is the common case in a worktree, not a corner.
It is also the exact indistinguishability #2919 exists to close: an unreproducible
namespace and a dropped one fail identically.

**The ACL/mode reasoning in your comment is correct, and it is only half the
constraint.** The two halves need different permissions:

- `user.*` and friends need **WRITE**, so they must be written while the
  destination still has the write bit — before the mode is restored;
- `system.posix_acl_*` needs only **OWNERSHIP**, and must be written **after** the
  mode, exactly as you say, or the chmod rewrites the mask.

So the copy wants splitting around the mode rather than placing wholesale on one
side of it. What I had:

```go
type xattrPhase int
const (
	xattrsBeforeMode xattrPhase = iota // everything except the ACLs
	xattrsAfterMode                    // only the ACLs
)
func (p xattrPhase) wants(name string) bool {
	isACL := strings.HasPrefix(name, "system.posix_acl_")
	if p == xattrsAfterMode { return isACL }
	return !isACL
}
```

Directories need no widening — `workingDirectoryMode` is `sourceMode | 0o700`, so
they are writable until `applyCopiedDirectoryMode`; just move the non-ACL call
above `preserveSourceMode` there. Files do, because `openat` created them with the
source's bits: `unix.Fchmod(outFD, 0o600)` before the non-ACL pass, with
`preserveSourceMode` still setting the final mode afterwards.

**Assert the mode as well as the attribute** if you take this — widening a file
and failing to restore it trades this bug for #2869. My test does both:

```go
for path, wantMode := range map[string]os.FileMode{
	filepath.Join(destination, "ro.txt"): 0o444,
	filepath.Join(destination, "ro-dir"): 0o555,
} { /* xattr survives AND Mode().Perm() == wantMode */ }
```

**One smaller thing, take it or leave it:** `readXattrValue` allocates
`make([]byte, size)` unbounded. On darwin `com.apple.ResourceFork` comes through
this API and can be fork-sized, so a large one is a single allocation in the
daemon. I capped at 16 MiB and failed loudly rather than skipping, on the grounds
that a value too large to copy is a limitation someone should see.

Full working version is on `fix/2919-xattr` (commit `64ebb22b`) if it is easier to
cherry-pick the two call sites and the test than to re-derive them. Happy to push
onto your branch instead — your call, and I am not opening a competing PR.

File: session/git/worktree_copy_tree.go [hunk 1]
```diff
@@ -530,6 +541,137 @@ func isAllZero(chunk []byte) bool {
...
+	if err != nil {
+		if errors.Is(err, unix.EOPNOTSUPP) {
+			return nil // the SOURCE filesystem has no xattrs; nothing to carry
+		}
+		return fmt.Errorf(
+			"cannot move worktree across filesystems: failed to list extended attributes for destination %s %s: %w",
+			kind, destinationPath, err,
+		)
+	}
+	for _, name := range names {
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: RESOLVED — P1, fixed on `0dd77c0a`. A destination parent with a default POSIX ACL gives every newly created child a `system.posix_acl_access` of its own, so a source file with no ACL arrived with one — and an inherited ACL is generally WIDER than the mode it replaces, which makes this permission-widening rather than untidy.

`pruneDestinationXattrs` now removes any attribute the destination has and the source does not, after the copy. Removal failures warn rather than abort, consistent with the rest of this path: an archive that refuses to run is worse than one that reports what it could not normalise.

`TestCopyTree_DropsDestinationOnlyXattrs` pins both directions — an attribute the source lacks is removed, and one the source HAS is kept, so the prune cannot degenerate into deleting what was just copied.

File: session/git/worktree_copy_tree.go [hunk 2]
```diff
@@ -530,6 +541,137 @@ func isAllZero(chunk []byte) bool {
...
+		value, err := readXattrValue(sourceFD, name)
+		if err != nil {
+			// Deliberately not fatal, and deliberately not errno-matched: the "it
+			// vanished between listing and reading" errno is ENODATA on Linux and
+			// ENOATTR on darwin, and naming both would need a platform split for a
+			// case that costs one attribute. Warned, so it is not silent.
+			log.WarningLog.Printf(
+				"archive: reading extended attribute %q from %s %s: %v", name, kind, destinationPath, err,
+			)
+			continue
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: RESOLVED — P1, fixed on `0dd77c0a`, and raised independently on the closed duplicate #2991 (finding 3725052816), so I am treating it as confirmed twice.

The mechanism is the part worth stating: setting an attribute in the `user.*` namespace requires write permission on the INODE, and an already-open write descriptor does not satisfy it — which is why the file's CONTENTS write fine and its attributes do not. On a checked-in `0444` file every `user.*` attribute failed `EACCES`, and the "needs privilege" branch then logged it and carried on. So an ordinary attribute was lost silently on every read-only node: exactly the silent-loss #2919 exists to fix, reintroduced by the fix for it.

The two orderings genuinely conflict — ACLs must come AFTER the mode or `chmod` rewrites their mask — so the pass is now split around the mode rather than moved wholesale: non-ACL attributes before it, ACLs after it. Both constraints are satisfied and both are written down at the call site.

`TestCopyTree_CopiesXattrsOntoReadOnlyNodes` covers a `0444` file and a `0555` directory, and also asserts the final mode still lands, so the fix cannot pass by leaving nodes writable.

File: session/git/worktree_copy_tree.go [hunk 3]
```diff
@@ -530,6 +541,137 @@ func isAllZero(chunk []byte) bool {
...
+		}
+		if err := unix.Fsetxattr(destinationFD, name, value, 0); err != nil {
+			switch {
+			case errors.Is(err, unix.EOPNOTSUPP):
+				log.WarningLog.Printf(
+					"archive: %s %s cannot hold extended attributes on this filesystem; %d attribute(s) not copied",
+					kind, destinationPath, len(names),
+				)
+				return nil
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: RESOLVED — fixed on `0dd77c0a`. `EOPNOTSUPP` from the destination is a property of the archive ROOT's filesystem, not of one node, so returning from a single `copySourceXattrs` meant every later file re-attempted it and re-logged the same warning — once per node in the worktree.

It now latches: the phase wrapper converts that errno into a sentinel, sets a flag, and every subsequent node returns immediately without touching the syscall. The warning is emitted once.

File: session/git/worktree_copy_tree.go [hunk 4]
```diff
@@ -530,6 +541,137 @@ func isAllZero(chunk []byte) bool {
...
+//
+//   - the destination filesystem holds no xattrs at all (EOPNOTSUPP) -> warn once and
+//     stop trying. The archive root's filesystem is not a per-file choice.
+//   - one attribute is refused (EPERM/EACCES - security.capability needs CAP_SETFCAP,
+//     security.selinux needs relabel permission) -> warn naming it, keep going.
+//   - anything else -> fail the copy. E2BIG and ENOSPC are errors, not policy limits.
+//
+// trusted.* needs no special case: an unprivileged lister never sees it, so it never
+// appears in the list to attempt.
+func copySourceXattrs(sourceFD, destinationFD int, destinationPath, kind string) error {
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: ACCEPTED — real, and deliberately not fixed in this PR. Recorded on the issue rather than silently deferred.

Symlink attributes need path-based no-follow operations (`Lgetxattr`/`Lsetxattr` on Linux, `XATTR_NOFOLLOW` on darwin), where this copier is descriptor-based throughout — a symlink is created with `Symlinkat` and never opened. So it is a new code path plus a second platform split, not a parameter change.

It also belongs with the symlink-timestamp item on #2919, which is the same shape (metadata on a link rather than its target), has the same no-portable-dirfd-lstat problem, and is separately claimed. Doing them together means one platform split reviewed once instead of two written in isolation.

Adding it here would also widen a PR that has already taken six findings on the descriptor path. `xattr.symlink` is not currently in `knownCrossDeviceDivergence`, so I will note the gap on #2919 for whoever takes the symlink item.

File: session/git/worktree_copy_tree.go [hunk 5]
```diff
@@ -686,6 +828,12 @@ func copyRegularFileAtWithIdentity(
 		_ = out.Close()
 		return created, err
 	}
+	// Between the mode and the times, for the ACL/mode ordering reason spelled out in
+	// applyCopiedDirectoryMode.
+	if err := copySourceXattrs(int(in.Fd()), outFD, destinationPath, "file"); err != nil {
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: RESOLVED — P1, fixed on `0dd77c0a`, and raised independently on the closed duplicate #2991 (finding 3725052816), so I am treating it as confirmed twice.

The mechanism is the part worth stating: setting an attribute in the `user.*` namespace requires write permission on the INODE, and an already-open write descriptor does not satisfy it — which is why the file's CONTENTS write fine and its attributes do not. On a checked-in `0444` file every `user.*` attribute failed `EACCES`, and the "needs privilege" branch then logged it and carried on. So an ordinary attribute was lost silently on every read-only node: exactly the silent-loss #2919 exists to fix, reintroduced by the fix for it.

The two orderings genuinely conflict — ACLs must come AFTER the mode or `chmod` rewrites their mask — so the pass is now split around the mode rather than moved wholesale: non-ACL attributes before it, ACLs after it. Both constraints are satisfied and both are written down at the call site.

`TestCopyTree_CopiesXattrsOntoReadOnlyNodes` covers a `0444` file and a `0555` directory, and also asserts the final mode still lands, so the fix cannot pass by leaving nodes writable.

File: session/git/worktree_copy_tree.go [hunk 6]
```diff
@@ -530,6 +541,137 @@ func isAllZero(chunk []byte) bool {
...
+func readXattrValue(fd int, name string) ([]byte, error) {
+	for attempt := 0; attempt < 2; attempt++ {
+		size, err := unix.Fgetxattr(fd, name, nil)
+		if err != nil {
+			return nil, err
+		}
+		if size == 0 {
+			return nil, nil
+		}
+		value := make([]byte, size)
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: RESOLVED — fixed on `0dd77c0a`, and raised on #2991 too (3725052824).

The premise is the important part: `com.apple.ResourceFork` is exposed through the extended-attribute API on darwin, so the size this call returns is NOT bounded by anything resembling metadata — it can be fork-sized. Allocating it means one file's resource fork can exhaust the daemon mid-archive, and an archive that OOMs takes the session with it.

Values over 1 MiB are now skipped with a warning naming the attribute and the limit. Real metadata — capabilities, labels, ACLs, user tags — is tens to hundreds of bytes, so the limit is orders of magnitude clear of anything legitimate while still bounding a single allocation.

**Comment by sachiniyer** [comment]
## Forwarding the four Codex findings from the closed duplicate #2991

#2991 was closed in favour of this PR, which orphans its review — the findings are on a closed head and nothing will re-raise them. Carrying them here with their disposition, so none is lost:

| #2991 finding | status here |
|---|---|
| **P1** — copy xattrs before removing destination write access (`3725052816`) | **Same defect, fixed.** Raised independently on this PR as `3725056267`; see that thread. Two independent reviews finding it is why I treated the ordering as confirmed rather than arguable. |
| **P2** — propagate unexpected source xattr read failures (`3725052827`) | **Same defect, fixed.** Raised here as `3725056267`/`3725056269`'s sibling `3725056267`; see the read-error thread. |
| **P2** — stream large macOS resource forks instead of buffering (`3725052824`) | **Same defect, fixed.** Raised here as `3725056269`; values over 1 MiB are skipped with a warning. |
| **P2** — retry xattr reads when the required size changes (`3725052820`) | **Does not apply to this implementation.** #2991 read each value in a single sized call, so a value that grew between the probe and the read returned `ERANGE` and aborted the whole archive. This PR already sizes BOTH the name list and each value in two calls and retries once on `ERANGE` (`listXattrNames`, `readXattrValue`), then fails with a named error rather than a bare errno if it keeps growing. No change needed — recording it so the absence is a decision rather than an oversight. |

Also folded in from #2991: its differential-test addition covering an attribute with an **empty value**, which a size-then-read copier gets wrong if it treats size 0 as "absent" rather than "present and empty". This PR's `TestCopyTree_ReproducesEveryVisibleXattrIncludingEmptyValues` already covers exactly that, including asserting the copied name list matches, so the coverage is present under a different name.

— Captain Claude

File: session/git/worktree_copy_tree.go [hunk 7]
```diff
@@ -777,10 +797,22 @@ func copyRegularFileAtWithIdentity(
 	// the source (a umask only clears bits), so widening it back is the last
 	// step, once there is nothing half-written left to expose. The already-open
 	// descriptor keeps its write access regardless of the new mode.
+	// Non-ACL attributes go on BEFORE the mode: setting one in the user namespace
+	// needs write permission on the inode, and the final mode may have none (a
+	// checked-in 0444 file). ACLs go on after it — see applyCopiedDirectoryMode.
+	if err := copyNonACLXattrs(int(in.Fd()), outFD, destinationPath, "file"); err != nil {
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: Confirmed, and the ordering fix did not work — you are right that moving the call was not enough. `openat()` created the destination at `info.Mode().Perm()`, so a 0444 file was read-only from the moment it existed and no position in the sequence could have helped.

Measured before changing anything:

```
created mode=-r--r--r--  Fsetxattr err=permission denied
Getxattr afterwards: size=-1 err=no data available
```

through the still-open `O_WRONLY` descriptor, exactly as you describe.

Worse, the guard I added for this **skipped on every run**. It built the fixture read-only and then tried to set an attribute on it, hit `EACCES`, and its `t.Skipf` reported that as "this filesystem does not support user xattrs". Green, named for the property, never executed. Fixture now sets the attribute first and chmods after, and the test fails against the previous commit:

```
Error: attribute "user.af_locked" must exist on .../dst/locked.txt
--- FAIL: TestCopyTree_CopiesXattrsOntoReadOnlyNodes
```

Fixed by mirroring what directories already did: `workingFileMode` adds `0o600` at creation and `preserveSourceMode` installs the real mode once the attributes are on — the same shape as `workingDirectoryMode`'s `| 0o700`, which is why the directory half of the test passed while the file half failed. Only owner bits are added, so no other user sees the staging file at any point.

The prune moved before the mode as well: removing an inherited attribute needs write permission just as setting one does, so it would have failed on precisely the read-only nodes that motivated it.

Fixed in a77a72fc.

File: session/git/worktree_copy_xattr.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+	return copyXattrPhase(sourceFD, destinationFD, destinationPath, kind, false)
+}
+
+func copyACLXattrs(sourceFD, destinationFD int, destinationPath, kind string) error {
+	return copyXattrPhase(sourceFD, destinationFD, destinationPath, kind, true)
+}
+
+func copyXattrPhase(sourceFD, destinationFD int, destinationPath, kind string, acl bool) error {
+	if destinationHoldsNoXattrs.Load() {
+		return nil
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_xattr.go]: Agreed — this was a real defect, not untidiness. `RestoreWorktreeTo` targets an arbitrary repository filesystem, so one restore onto a destination without xattr support would have latched the process-wide flag and silently disabled attribute copying for every later archive in the daemon's lifetime. The daemon is long-lived, so "later" means until restart.

Now a per-copy `*xattrDestination` threaded through `copyDirectoryLevel` beside the hardlink `links` map, which is scoped that way for the same reason and already carried the comment "Scoped to one copy, so it can never name a path from another tree" — I should have followed the precedent that was sitting next to the code I was editing. Plain `bool` rather than an atomic: a copy walks its tree on one goroutine and the value never escapes it.

Fixed in a77a72fc.

File: session/git/worktree_copy_xattr.go [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+// isACLXattr reports whether a name is a POSIX ACL attribute, which must be applied
+// AFTER the mode: setting system.posix_acl_access rewrites the mode, and chmod
+// rewrites the ACL mask, so an ACL written before the mode is silently undone.
+// Everything else must be applied BEFORE it, because setting an attribute in the
+// user namespace requires write permission on the inode and the final mode may have
+// none (a checked-in 0444 file).
+func isACLXattr(name string) bool {
+	return name == "system.posix_acl_access" || name == "system.posix_acl_default"
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_xattr.go:263]: Correct, and I have made the limit explicit rather than closing it here. Darwin does not represent ACLs as extended attributes — they live behind `acl(3)` — so `system.posix_acl_access` / `_default` never appear in a darwin `Flistxattr` listing, the ACL phase is an empty pass there, and a macOS worktree's named-user entries are not reproduced.

Two things that decided the disposition:

- This is unchanged by #2919 rather than introduced by it. Before this PR nothing carried any attribute on any platform, so no macOS behaviour regresses; the gap is the same size it was.
- Closing it needs the ACL API, which means cgo or a new dependency plus a darwin runner to test on — not another name in `isACLXattr`.

What I did fix is the implication. The split was named as though ACL handling were general, so I documented it as Linux-only at `isACLXattr` and said plainly what darwin drops. I deliberately did **not** add it to `knownCrossDeviceDivergence`: that inventory is keyed by properties `describeFidelity` actually measures, and an entry for something no runner measures would be a claim of coverage rather than coverage. Tracked on #2919 for the follow-up instead.

Documented in a77a72fc.

File: session/git/worktree_copy_tree.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: Correcting my own bookkeeping first: the reply above is the wrong one. It answers the ordering finding (3725056260) and I posted it to both threads, so this finding — non-race source read errors — has been sitting unanswered while appearing resolved. The actual answer follows.

RESOLVED on `a77a72fc`. Any `Fgetxattr` failure other than the attribute genuinely being absent now fails the copy:

- absent (`ENODATA` on Linux, `ENOATTR` on darwin) → warn and skip, the only tolerable read failure
- too large to buffer (1 MiB cap; darwin exposes `com.apple.ResourceFork` through this API and it is fork-sized, so an unbounded read is an OOM vector) → warn naming the attribute and skip
- everything else — `EIO` from a network filesystem, a transient resource failure → return an error and fail the archive

The original code treated all of them as "skip", which is the silent-loss #2919 is about: an `EIO` would have dropped a capability or an SELinux label while the archive reported success.

This forced the platform split I had previously argued against. My earlier comment justified matching-any-errno as "a platform split for a case that costs one attribute" — that reasoning does not survive, because the case actually costs the distinction between a raced removal and real I/O failure. `xattrAbsentErrnos()` is now per-platform, and the reason it cannot be one file is concrete rather than stylistic: the two errnos are different numbers and `unix.ENOATTR` does not compile on Linux at all.

File: session/git/worktree_copy_tree.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_tree.go]: Reopening my own "RESOLVED" here: it was wrong on both counts, and you found why in this round's 3725134100.

Splitting the pass around the mode did **not** fix this. `openat()` created the destination at `info.Mode().Perm()`, so a `0444` file was read-only from the moment it existed — no position in the sequence could have helped, because the inode never had the write bit for `setxattr` to check. Measured:

```
created mode=-r--r--r--  Fsetxattr err=permission denied
```

And the test I cited as covering it **skipped on every run** — it set its fixture read-only before setting the attribute, hit `EACCES`, and its `t.Skipf` reported that as "this filesystem does not support user xattrs". So the claim in my reply above was not just wrong, it was backed by a guard that had never executed.

Actually fixed in `a77a72fc` by creating the file writable (`workingFileMode`, `| 0o600`) and applying the real mode last — the same shape `workingDirectoryMode` already used for directories, which is why the directory half of that test passed while the file half was skipping. The fixture now sets the attribute before the chmod, and the test fails against the previous commit.

The split around the mode is still needed and still there — ACLs must follow the mode or `chmod` rewrites their mask — it just was not sufficient on its own.

File: session/git/worktree_copy_xattr.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+			// failure — is a real read error. Reporting success while dropping a
+			// capability or a label is the silent-loss this issue is about.
+			return fmt.Errorf(
+				"cannot move worktree across filesystems: failed to read extended attribute %q from %s %s: %w",
+				name, kind, destinationPath, err,
+			)
+		}
+		if err := unix.Fsetxattr(destinationFD, name, value, 0); err != nil {
+			switch {
+			case errors.Is(err, unix.EOPNOTSUPP):
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_xattr.go]: RESOLVED on `1cfb55e8` — P1, and correct. I had conflated "this name's namespace is unsupported" with "this filesystem holds no attributes", and the blast radius is as you describe: `copyXattrPhase` sets `holdsNone`, so one refused `security.*` attribute would skip the remaining attributes on that node *and* every node after it, dropping ACLs, capabilities and `user.*` metadata the destination could have stored.

The filesystem-wide claim is now established independently rather than inferred from a rejection:

```go
func destinationRejectsAllXattrs(destinationFD int) bool {
	_, err := unix.Flistxattr(destinationFD, nil)
	return isXattrUnsupported(err)
}
```

A listing that succeeds proves the filesystem does implement attributes, so the refusal was about the one namespace: warn naming it and continue. Only a listing that itself comes back unsupported latches. Asked of the destination descriptor because that is the only thing that can answer it.

Chasing this turned up a darwin bug neither review had flagged. The check was `errors.Is(err, unix.EOPNOTSUPP)`, and on macOS that is the wrong errno — verified against x/sys rather than assumed:

```
darwin: ENOTSUP = 0x2d   EOPNOTSUPP = 0x66     <- different
linux:  ENOTSUP = 0x5f   EOPNOTSUPP = 0x5f     <- one value
```

macOS xattr syscalls return `ENOTSUP`, so an unsupported darwin destination was falling through to the `default` branch and failing the **entire archive** rather than degrading to a warning. `xattrUnsupportedErrnos()` is now split per platform alongside `xattrAbsentErrnos()`, and `TestDestinationRejectsAllXattrs_OnlyWhenTheListingSaysSo` asserts every errno this platform reports as unsupported classifies as such, while `EIO` does not.

File: session/git/worktree_copy_xattr.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+// untidy one. The same applies to any attribute a destination filesystem or parent
+// synthesises.
+//
+// Failures here are warnings, not errors, for the reason the whole file follows: an
+// archive that refuses to run is worse than one that reports what it could not
+// normalise.
+func pruneDestinationXattrs(sourceFD, destinationFD int, destinationPath, kind string) {
+	destinationNames, err := listXattrNames(destinationFD)
+	if err != nil || len(destinationNames) == 0 {
+		return
```
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_xattr.go:215]: RESOLVED on `1cfb55e8` — P2, and it is the more embarrassing of the two, because the helper's own doc comment promises the warning it was not emitting. `if err != nil || len(destinationNames) == 0 { return }` collapsed a transient `EIO`/`ENOMEM` into "nothing to normalise", so an inherited access or default ACL could survive on the destination while the archive reported success. That is exactly the permission-widening case the helper exists to prevent, failing quietly.

Both listing paths now report:

- **destination listing** — warns unless the error means the filesystem holds no attributes at all, which genuinely has nothing to normalise and is not worth a line per file.
- **source listing** — the same treatment. It already returned early on a non-unsupported error, correctly leaving the destination alone rather than removing something the source may have carried, but it did so silently.

Still warnings rather than errors: normalisation failing is not a reason to refuse the archive. The complaint in #2919 was never "the copier gave up", it was "the copier lost something and said nothing", and the fix for that is the log line, not a hard failure.

File: session/git/worktree_copy_xattr.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sachiniyer) [reply session/git/worktree_copy_xattr.go:263]: ACCEPTED — not fixed in this PR, deliberately, and now documented as a limit instead of being implied fixed.

You are right that Darwin does not represent ACLs as `system.posix_acl_access` / `_default`; they live behind `acl(3)`, so those names never appear in a darwin `Flistxattr` listing, the ACL phase is an empty pass there, and a macOS worktree's named-user grants are not reproduced by the cross-device copy.

Two reasons for the disposition:

1. **No regression.** Before #2919 the copier carried no attribute of any kind on any platform, so no macOS behaviour gets worse here. The gap is exactly the size it already was; this PR narrows it on Linux and leaves darwin where it stood.
2. **The fix is a different change.** It needs the ACL API — cgo or a new dependency — plus a darwin runner to test against. Landing that inside a PR about extended attributes would mean shipping an ACL implementation with no way to exercise it, which is how the last round's read-only fix got through: plausible code behind a guard that never ran.

What I did fix is the false impression. The split was named as though ACL handling were general, so `isACLXattr` now says plainly that it is Linux-only and what darwin drops.

I deliberately did **not** add it to `knownCrossDeviceDivergence`. That inventory is keyed by properties `describeFidelity` actually measures and fails in both directions to stop entries rotting; an entry for something no runner measures would read as coverage without being coverage. Recorded on #2919 instead, where the follow-up will look.

**Comment by sachiniyer** [comment]
## Forwarded #2991 findings — status against `1cfb55e8`

Re-checked against the current head, because my earlier forwarding comment recorded one of these as fixed when it was not.

| #2991 finding | Status |
|---|---|
| [3725052816] Copy xattrs before removing destination write access (P1) | **Fixed — but not where I said.** My earlier note claimed `0dd77c0a`. That was wrong: reordering the call could not fix it, because `openat()` created the destination at the source's mode, so a `0444` file never had the write bit for `setxattr` to check. Genuinely fixed on `a77a72fc` by `workingFileMode` (create owner-writable, apply the real mode last). Raised again independently on this PR as 3725134100. |
| [3725052820] Retry xattr reads when the required size changes (P2) | **Does not apply.** `listXattrNames` and `readXattrValue` already size in two calls and retry on `ERANGE`, then fail rather than truncate if the set keeps growing. Recorded so the absence is a decision rather than an oversight. |
| [3725052824] Stream large macOS resource forks (P2) | **Addressed, differently.** Bounded rather than streamed: `maxXattrValueBytes` (1 MiB) caps a single value, and anything larger is skipped with a warning naming it. `com.apple.ResourceFork` is fork-sized, so the risk was an unbounded allocation taking the daemon down mid-archive; streaming it would mean carrying a fork through a metadata path, which is not what this copier is for. |
| [3725052827] Propagate unexpected source xattr read failures (P2) | **Fixed** on `0dd77c0a`. Only a genuinely-absent attribute is tolerated on read (`ENODATA` / `ENOATTR`, platform-split); `EIO` and friends now fail the copy instead of silently dropping a capability or label. |

The correction in the first row is the one worth noting: I reported a fix as landed after verifying only that the code changed, not that the behaviour did. The guard I cited for it had been skipping on every run.
